### ISSUE TITLE
Proposal: Report warnings for non-public Fact attributes

### ISSUE THREAD

**Issue by nxtn** [xunit/xunit#2120]
* Is this a bug with VSTest (aka Visual Studio Test Explorer, `dotnet test`, `dotnet vstest`, or `vstest.console.exe`)?

  Maybe, but I consider this as an *analyzer* issue.

xUnit ignores test methods with non-public custom `Fact` attributes and other methods in the same class. If this is intentional, the analyzer should be updated to report warnings about such attributes.

## Version

xunit 2.4.1
xunit.runner.visualstudio 2.4.2

## Steps to reproduce

```cs
using System;
using Xunit;

[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
internal sealed class CustomFactAttribute : FactAttribute
{
    internal CustomFactAttribute()
    {
        this.Skip = "xxx";
    }
}

public class Tests
{
    [CustomFact]
    public void TestCustomFact()
    {
    }

    [Fact]
    public void TestFact()
    {
    }
}
```

## Expected

```
Total tests: 2
     Passed: 1
    Skipped: 1
```

## Actual

```
No test is available in xxx. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
```

**Comment by GeraldLx** [comment]
Perhaps I miss something, but your problem seems not related to the custom attribute.

In your sample code you say Skip = "xxx", basically telling Xunit not to run the tests.

You will get the same result when you just write:
[Fact(Skip = "xxx")]

**Comment by NextTurn** [comment]
> Perhaps I miss something, but your problem seems not related to the custom attribute.
> 
> In your sample code you say Skip = "xxx", basically telling Xunit not to run the tests.
> 
> You will get the same result when you just write:
> [Fact(Skip = "xxx")]

If I write `[Fact(Skip = "xxx")]` or `public CustomFactAttribute()`, I will get the "Expected" results.

Note that I only need to change the accessibility of the *constructor* rather than the *type*.

**Comment by GeraldLx** [comment]
Oh, you are right, I made a similiar project with an internal attribute and could not find any errors, because my constructor was public.

So you are right that this will not work if your attribute does not have any public constructors, but I would not say that this is a Xunit bug. E.g. almost every Container framework I know works the same.

So basically you tell Xunit to use an attribute, but hiding how to build it.

This is e.g. what my test runner outputs:
An error occurred while analysing this project after it was built: System.Exception: Exception during discovery:
System.MissingMethodException: Constructor on type 'ClassLibrary18.CustomFactAttribute' not found.

I see good reasons e.g. to make an attribute internal, but why do you want the constructor also internal?

**Comment by NextTurn** [comment]
I didn't consider this as a bug. xUnit has an analyzer that reports warnings for stupid things like not making test classes public or not placing `[Fact]` on test methods, and I hope it can be improved to include this.

> I see good reasons e.g. to make an attribute internal, but why do you want the constructor also internal?

It was just an accident, I think.

**Comment by bradwilson** [comment]
I agree, this should be an analyzer, so I'm marking it as such.

**Comment by JamesTerwilliger** [comment]
I put together an analyzer for this in this PR: https://github.com/xunit/xunit.analyzers/pull/175

However, I don't think it's quite complete or quite right. Its current form identifies that if a class is a subclass of FactAttribute, then any constructor that is internal or protected internal should be public. And it analyzes the class definition, not the references.

If the analysis stays on the class itself, then the question is whether this is the right set of constructors to make public. Should only the default/0-parameter constructor need to be public? Do they all need to be public? Does at least one need to be public?

Alternatively, I could easily move the analysis to the call sites/attribute references themselves. That would ensure that only constructors that are actually used in an Attribute context are flagged, but it would make a fixer fairly impossible or at least unlikely as the class may be defined outside the analysis context.

Thoughts welcome.

**Comment by bradwilson** [comment]
James and I discussed this in the PR, and the logic will be that internal constructors are allowed, so long as they aren't used from the test method. Since this causes runtime failures, this is being flagged as an Error.

Available in `1.10.0-pre.9` https://xunit.net/docs/using-ci-builds
